experimental/yarpl/src/yarpl/flowable/sources/Flowable_FromObservable.h

```diff
@@ -60,97 +50,57 @@ class FlowableFromObservableSubscription
-      tryDelete();
+      observableSubscription_->cancel();
+      release();
     }
   }
 
- private:
-  std::shared_ptr<yarpl::observable::Observable<T>> observable_;
-  std::unique_ptr<reactivestreams_yarpl::Subscriber<T>> subscriber_;
-  std::atomic_bool started{false};
-  std::atomic<int64_t> requested_{0};
-
-  void tryDelete() {
-    // only one thread can call delete
-
-    // TODO when is it safe to delete?
-    //    if (emitting_.fetch_add(1) == 0) {
-    //      // TODO remove this cout once happy with it
-    //      std::cout << "Delete FlowableSubscriptionSync" << std::endl;
-    //      delete this;
-    //    }
+  // Observer override
+  void onSubscribe(
+      Reference<yarpl::observable::Subscription> subscription) override {
+    observableSubscription_ = subscription;
   }
 
-  void onNext(const T& t) {
+  // Observer override
+  void onNext(const T& t) override {
+    if (++sent % 1000000 == 0) {
+    }
     if (requested_ > 0) {
       subscriber_->onNext(t);
       yarpl::flowable::internal::SubscriptionHelper::consumeCredits(
           &requested_, 1);
     }
-    // drop anything else
+    // drop anything else received while we don't have credits
   }
 
-  void onComplete() {
+  // Observer override
+  void onComplete() override {
     subscriber_->onComplete();
+    release();
   }
 
-  void onError(const std::exception_ptr error) {
+  // Observer override
+  void onError(const std::exception_ptr error) override {
     subscriber_->onError(error);
+    release();
   }
 
-  /**
-   * This will be used to subscribe to the upstream Observable.
-   *
-   * The FlowableFromObservableSubscription* contained by it will live
-   * until we send terminal events or are cancelled.
-   *
-   * NOTE on lifecycle: This class can be deleted BEFORE
-   * the Flowable which is represented by the subscription above.
-   * This would happen if the Observable sends an onComplete/onError,
-   * so we MUST not assume TheObserver will exist and retain
-   * references to it.
-   */
-  class TheObserver : public yarpl::observable::Observer<T> {
-   public:
-    explicit TheObserver(FlowableFromObservableSubscription<T>* fs) : fs_(fs) {}
-
-    void onSubscribe(yarpl::observable::Subscription* os) override {
-      os_ = os;
-    }
-
-    void onNext(const T& t) override {
-      fs_->onNext(t);
-    }
-
-    void onComplete() override {
-      fs_->onComplete();
-    }
-
-    void onError(const std::exception_ptr error) override {
-      fs_->onError(error);
-    }
-
-    /**
-     * Allow the Flowable Subscriber to cancel the Observable.
-     */
-    void cancel() {
-      os_->cancel();
-    }
-
-   private:
-    FlowableFromObservableSubscription<T>* fs_;
-    yarpl::observable::Subscription* os_;
-  };
+ private:
+  Reference<yarpl::observable::Observable<T>> observable_;
+  Reference<yarpl::flowable::Subscriber<T>> subscriber_;
+  std::atomic_bool started{false};
+  std::atomic<int64_t> requested_{0};
+  class TheObserver;
```

Comment:
Yup it sure was. So much for fully reviewing the diff before submitting it. Thanks for the catch.